src/azstoragetorch/_client.py

```diff
@@ -0,0 +1,146 @@
+    _RETRYABLE_READ_EXCEPTIONS = (
+        azure.core.exceptions.IncompleteReadError,
+        azure.core.exceptions.HttpResponseError,
+        azure.core.exceptions.DecodeError,
+    )
+
+    def __init__(
+        self,
+        sdk_blob_client: azure.storage.blob.BlobClient,
+        executor: Optional[concurrent.futures.Executor] = None,
+    ):
+        self._sdk_blob_client = sdk_blob_client
+        self._generated_sdk_storage_client = self._sdk_blob_client._client
+        if executor is None:
+            executor = concurrent.futures.ThreadPoolExecutor()
+        self._executor = executor
+
+    @classmethod
+    def from_blob_url(
+        cls, blob_url: str, sdk_credential: SDK_CREDENTIAL_TYPE
+    ) -> "AzStorageTorchBlobClient":
+        sdk_blob_client = azure.storage.blob.BlobClient.from_blob_url(
+            blob_url,
+            credential=sdk_credential,
+            connection_data_block_size=cls._CONNECTION_DATA_BLOCK_SIZE,
+        )
+        return AzStorageTorchBlobClient(sdk_blob_client)
+
+    def get_blob_size(self) -> int:
+        return self._blob_properties.size
+
+    def download(self, offset: int = 0, length: Optional[int] = None) -> bytes:
+        length = self._update_download_length_from_blob_size(offset, length)
+        if length < self._PARTITIONED_DOWNLOAD_THRESHOLD:
+            return self._download_with_retries(offset, length)
+        else:
+            return self._partitioned_download(offset, length)
+
+    def close(self) -> None:
+        self._executor.shutdown()
+
+    @functools.cached_property
+    def _blob_properties(self) -> azure.storage.blob.BlobProperties:
+        return self._sdk_blob_client.get_blob_properties()
+
+    def _update_download_length_from_blob_size(
+        self, offset: int, length: Optional[int] = None
+    ) -> int:
+        length_from_offset = self.get_blob_size() - offset
+        if length is not None:
+            return min(length, length_from_offset)
+        return length_from_offset
+
+    def _partitioned_download(self, offset: int, length: int) -> bytes:
+        futures = []
+        for partition in self._get_partitioned_reads(offset, length):
+            futures.append(
+                self._executor.submit(self._download_with_retries, *partition)
+            )
+        return b"".join(f.result() for f in futures)
+
+    def _get_partitioned_reads(self, offset: int, length: int) -> List[Tuple[int, int]]:
+        end = offset + length
+        num_partitions = math.ceil(length / self._PARTITION_SIZE)
+        partitions = []
+        for i in range(num_partitions):
+            start = offset + i * self._PARTITION_SIZE
+            if start >= end:
+                break
+            size = min(self._PARTITION_SIZE, end - start)
+            partitions.append((start, size))
+        return partitions
+
+    def _download_with_retries(self, pos: int, length: int) -> bytes:
+        attempts_remaining = 3
+        while attempts_remaining > 0:
```

Comment:
We should probably add some sleep time in between tries. Retrying without sleep may not help and may exasxcerbate network issues.

Reply:
I added exponential backoff with full jitter as a starting point here. Mainly it will at least pause and slightly stagger streaming across workers. In the future, if connection issues become an issue especially because of overwhelming the network, we should consider adapting how fast we read from the stream as an approach to reducing the network errors. I added this as a comment as well to help remind us if we revisit it.